contracts/strategy/TradeFintechStrategy.sol

```diff
@@ -0,0 +1,134 @@
+            }
+        }
+        _totalAssets = totalInvestedInRWA + asset.balanceOf(address(this));
+    }
+
+    function availableDepositLimit(address /*_owner*/) public view override returns (uint256) {
+        // Return the remaining room.
+        return type(uint256).max - asset.balanceOf(address(this));
+    }
+
+    function availableWithdrawLimit(address /*_owner*/) public view override returns (uint256) {
+        return TokenizedStrategy.totalIdle() + totalInvestedInRWA;
+    }
+
+    function _deployFunds(uint256 _amount) internal override {
+        require(block.timestamp <= depositPeriodEnds, "Deposit period has ended");
+        asset.transfer(managerAddress, _amount);
+        totalInvestedInRWA += _amount;
+    }
+
+    function _freeFunds(uint256 _amount) internal override {
+        require(block.timestamp > lockPeriodEnds, "Lock period has not ended");
+        // We don't check available liquidity because we need the tx to
+        // revert if there is not enough liquidity so we don't improperly
+        // pass a loss on to the user withdrawing.
+        asset.safeTransferFrom(managerAddress, address(this), _amount);
+        totalInvestedInRWA -= _amount;
+    }
+
+    /**
+     * @dev Optional function for a strategist to override that will
+     * allow management to manually withdraw deployed funds from the
+     * yield source if a strategy is shutdown.
+     *
+     * This should attempt to free `_amount`, noting that `_amount` may
+     * be more than is currently deployed.
+     *
+     * NOTE: This will not realize any profits or losses. A separate
+     * {report} will be needed in order to record any profit/loss. If
+     * a report may need to be called after a shutdown it is important
+     * to check if the strategy is shutdown during {_harvestAndReport}
+     * so that it does not simply re-deploy all funds that had been freed.
+     *
+     * EX:
+     *   if(freeAsset > 0 && !TokenizedStrategy.isShutdown()) {
+     *       depositFunds...
+     *    }
+     *
+     * @param _amount The amount of asset to attempt to free.
+     */
+    function _emergencyWithdraw(uint256 _amount) internal override {
+        uint256 amountToWithdraw = Math.min(_amount, totalInvestedInRWA);
+        asset.safeTransferFrom(managerAddress, address(this), amountToWithdraw);
+        asset.transfer(TokenizedStrategy.management(), amountToWithdraw);
+        totalInvestedInRWA -= amountToWithdraw;
+    }
+
+
+    /// @notice Allows the manager to report gains or losses.
+    /// @dev Should be called before calling report() to report the amount of the gain or loss.
+    /// @dev The manager can only report gains or losses.
+    /// @param _gain The amount of the gain.
+    /// @param _loss The amount of the loss.
+    function reportGainOrLoss(uint256 _gain, uint256 _loss) external {
+        require(msg.sender == managerAddress, "Only the manager can report gains or losses");
+
+        if (_gain > 0) {
```

Comment:
Why there is no gain added to "totalInvestedInRWA" in case of gain, if there is sub for loss?

Reply:
Because in case of gain, on the next harvestAndReport trigger we will have the totalInvested updated.

In case of a loss, we have to update it immediately.

Reply:
When gains are made from investments managed by the smart contract, they are not immediately added to the totalInvested variable. This approach is chosen because gains are considered part of the natural fluctuation of the investment's value. Instead of directly adjusting the totalInvested every time a gain occurs, the contract waits until the next harvesting cycle. During this cycle, any "loose" assets (which include realized gains not yet accounted for) found in the contract are added to the totalInvested. This process ensures that the totalInvested accurately reflects the total value of assets the contract is managing after accounting for all fluctuations, including gains. If we do this also on the reportGainOrLoss we will be double counting the gain into the totalInvested.

On the other way, when a loss is reported, it's immediately subtracted from the totalInvested. This immediate action is necessary because a loss directly reduces the contract's asset base. Immediate adjustment ensures the totalInvested reflects the current, accurate value of assets after the loss. This distinction in handling gains and losses allows for more accurate and timely management of the contract's financial state, ensuring that losses are promptly accounted for, while gains are integrated during the routine financial reconciliation process (harvesting).

The critical distinction here is that losses directly reduce the principal (the amount initially invested), impacting the contract's immediate financial health and needing immediate adjustment in the totalInvested value. In contrast, gains increase the assets under management but do not impact the principal in the same direct way. They are realized and added to the principal during the next harvest cycle, ensuring that the total invested amount remains accurate over time without needing immediate adjustment upon the gain realization.

Reply:
Ok, I think I got what's the source of the misunderstanding here. You are thinking on reportGainOrLoss function as just a function really to be used to update the state. When the Strategy Manager reports a gain it won't necessarily transfer that gain back to the Strategy, right?

In this case, we need to update the totalInvested also on gain report. Will do this change.

Reply:
Just make sure there are no double gain if you make this change

Reply:
It is not possible to have double gain now because:

1. We have the totalInvested being increased upon gain reports on reportGainOrLoss
2. We have totalInvested being decreased upon the return of funds with returnFunds
3. On harvestAndReport we only increase totalInvested if any loose assets are transferred to the Strategy Manager